src/EditorFeatures/Core/Implementation/NavigationBar/NavigationBarController.cs

```diff
@@ -340,32 +328,46 @@ private void PushSelectedItemsToPresenter(NavigationBarSelectedTypeAndMember sel
             _presenter.PresentItems(
                 projectItems,
                 selectedProjectItem,
-                listOfLeft,
+                listOfLeft.ToImmutable(),
                 newLeft,
                 newRight);
-            _versionStampOfFullListPushedToPresenter = null;
         }
 
         private void OnItemSelected(object? sender, NavigationBarItemSelectedEventArgs e)
         {
             AssertIsForeground();
+            _ = OnItemSelectedAsync(e.Item);
+        }
 
-            _waitIndicator.Wait(
+        private async Task OnItemSelectedAsync(NavigationBarItem item)
+        {
+            AssertIsForeground();
+            using var waitContext = _waitIndicator.StartWait(
                 EditorFeaturesResources.Navigation_Bars,
                 EditorFeaturesResources.Refreshing_navigation_bars,
                 allowCancel: true,
-                action: context => ProcessItemSelectionSynchronously(e.Item, context.CancellationToken));
+                showProgress: false);
+
+            try
+            {
+                await ProcessItemSelectionAsync(item, waitContext.CancellationToken).ConfigureAwait(false);
+            }
+            catch (OperationCanceledException)
+            {
```

Comment:
cancelaltion through the twd is caught and doesn't bubble out to the task.  errors are handled below and will give us NFWs.

Reply:
I think this could just be FatalError.ReportAndCatchUnlessCanceled